# Data Collection and Initial Inspection

## Objectives

This notebook:

- documents the location and provenance of the project data;
- confirms that the required files are available;
- examines dataset dimensions and column names;
- investigates missing values and duplicated records;
- identifies the available date range;
- records initial data-quality observations.

No source data is modified in this notebook.

In [1]:
from pathlib import Path

import pandas as pd

PROJECT_ROOT = (
    Path.cwd().parent
    if Path.cwd().name == "jupyter_notebooks"
    else Path.cwd()
)

RAW_FOLDER = PROJECT_ROOT / "data" / "raw" / "v1"

GLOBAL_FILE = RAW_FOLDER / "GlobalTemperatures.csv"
COUNTRY_FILE = RAW_FOLDER / "GlobalLandTemperaturesByCountry.csv"

print(f"Project root: {PROJECT_ROOT}")
print(f"Global file: {GLOBAL_FILE}")
print(f"Country file: {COUNTRY_FILE}")

Project root: /Users/ewa/Documents/vscode-projects/code-institute-2026/CI-Project3/CI-Project3-ClimateLens-Understanding-Global-Temperature-Change
Global file: /Users/ewa/Documents/vscode-projects/code-institute-2026/CI-Project3/CI-Project3-ClimateLens-Understanding-Global-Temperature-Change/data/raw/v1/GlobalTemperatures.csv
Country file: /Users/ewa/Documents/vscode-projects/code-institute-2026/CI-Project3/CI-Project3-ClimateLens-Understanding-Global-Temperature-Change/data/raw/v1/GlobalLandTemperaturesByCountry.csv


In [2]:
assert GLOBAL_FILE.exists(), (
    f"Missing file: {GLOBAL_FILE}. "
    "Download it from Kaggle and place it in data/raw/v1."
)

assert COUNTRY_FILE.exists(), (
    f"Missing file: {COUNTRY_FILE}. "
    "Download it from Kaggle and place it in data/raw/v1."
)

print("Both required dataset files were found.")

Both required dataset files were found.


In [3]:
global_df = pd.read_csv(GLOBAL_FILE)
country_df = pd.read_csv(COUNTRY_FILE)

print(f"Global dataset shape: {global_df.shape}")
print(f"Country dataset shape: {country_df.shape}")

Global dataset shape: (3192, 9)
Country dataset shape: (577462, 4)


In [4]:
display(global_df.head())
display(country_df.head())

,dt,LandAverageTemperature,LandAverageTemperatureUncertainty,LandMaxTemperature,LandMaxTemperatureUncertainty,LandMinTemperature,LandMinTemperatureUncertainty,LandAndOceanAverageTemperature,LandAndOceanAverageTemperatureUncertainty
0,1750-01-01,3.034,3.574,NaN,NaN,NaN,NaN,NaN,NaN
1,1750-02-01,3.083,3.702,NaN,NaN,NaN,NaN,NaN,NaN
2,1750-03-01,5.626,3.076,NaN,NaN,NaN,NaN,NaN,NaN
3,1750-04-01,8.490,2.451,NaN,NaN,NaN,NaN,NaN,NaN
4,1750-05-01,11.573,2.072,NaN,NaN,NaN,NaN,NaN,NaN


,dt,AverageTemperature,AverageTemperatureUncertainty,Country
0,1743-11-01,4.384,2.294,Åland
1,1743-12-01,NaN,NaN,Åland
2,1744-01-01,NaN,NaN,Åland
3,1744-02-01,NaN,NaN,Åland
4,1744-03-01,NaN,NaN,Åland


In [5]:
expected_global_columns = {
    "dt",
    "LandAverageTemperature",
    "LandAverageTemperatureUncertainty",
    "LandMaxTemperature",
    "LandMaxTemperatureUncertainty",
    "LandMinTemperature",
    "LandMinTemperatureUncertainty",
    "LandAndOceanAverageTemperature",
    "LandAndOceanAverageTemperatureUncertainty",
}

expected_country_columns = {
    "dt",
    "AverageTemperature",
    "AverageTemperatureUncertainty",
    "Country",
}

missing_global_columns = expected_global_columns.difference(
    global_df.columns
)

missing_country_columns = expected_country_columns.difference(
    country_df.columns
)

assert not missing_global_columns, (
    f"Missing expected global columns: {missing_global_columns}"
)

assert not missing_country_columns, (
    f"Missing expected country columns: {missing_country_columns}"
)

print("All expected columns are present.")

All expected columns are present.


In [6]:
print("GLOBAL DATASET")
global_df.info()

print("\nCOUNTRY DATASET")
country_df.info()

GLOBAL DATASET
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3192 entries, 0 to 3191
Data columns (total 9 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   dt                                         3192 non-null   object 
 1   LandAverageTemperature                     3180 non-null   float64
 2   LandAverageTemperatureUncertainty          3180 non-null   float64
 3   LandMaxTemperature                         1992 non-null   float64
 4   LandMaxTemperatureUncertainty              1992 non-null   float64
 5   LandMinTemperature                         1992 non-null   float64
 6   LandMinTemperatureUncertainty              1992 non-null   float64
 7   LandAndOceanAverageTemperature             1992 non-null   float64
 8   LandAndOceanAverageTemperatureUncertainty  1992 non-null   float64
dtypes: float64(8), object(1)
memory usage: 224.6+ KB

COUNTRY DATASET
<class 'pandas.c

In [7]:
def missing_value_summary(dataframe):
    """Return the number and percentage of missing values per column."""
    summary = pd.DataFrame({
        "missing_count": dataframe.isna().sum(),
        "missing_percentage": (
            dataframe.isna().mean() * 100
        ).round(2),
    })

    return summary.sort_values(
        by="missing_percentage",
        ascending=False,
    )


print("Global dataset missing values")
display(missing_value_summary(global_df))

print("Country dataset missing values")
display(missing_value_summary(country_df))

Global dataset missing values


,missing_count,missing_percentage
LandMaxTemperature,1200,37.59
LandMaxTemperatureUncertainty,1200,37.59
LandMinTemperature,1200,37.59
LandMinTemperatureUncertainty,1200,37.59
LandAndOceanAverageTemperature,1200,37.59
LandAndOceanAverageTemperatureUncertainty,1200,37.59
LandAverageTemperature,12,0.38
LandAverageTemperatureUncertainty,12,0.38
dt,0,0.00


Country dataset missing values


,missing_count,missing_percentage
AverageTemperature,32651,5.65
AverageTemperatureUncertainty,31912,5.53
dt,0,0.00
Country,0,0.00


In [8]:
global_duplicate_dates = global_df.duplicated(
    subset=["dt"]
).sum()

country_duplicate_records = country_df.duplicated(
    subset=["Country", "dt"]
).sum()

print(f"Duplicated global dates: {global_duplicate_dates}")
print(
    "Duplicated country-and-date combinations: "
    f"{country_duplicate_records}"
)

Duplicated global dates: 0
Duplicated country-and-date combinations: 0


In [9]:
global_dates = pd.to_datetime(
    global_df["dt"],
    errors="coerce",
)

country_dates = pd.to_datetime(
    country_df["dt"],
    errors="coerce",
)

print("Global dataset")
print(f"Earliest date: {global_dates.min()}")
print(f"Latest date: {global_dates.max()}")
print(f"Invalid dates: {global_dates.isna().sum()}")

print("\nCountry dataset")
print(f"Earliest date: {country_dates.min()}")
print(f"Latest date: {country_dates.max()}")
print(f"Invalid dates: {country_dates.isna().sum()}")

Global dataset
Earliest date: 1750-01-01 00:00:00
Latest date: 2015-12-01 00:00:00
Invalid dates: 0

Country dataset
Earliest date: 1743-11-01 00:00:00
Latest date: 2013-09-01 00:00:00
Invalid dates: 0


In [10]:
country_counts = (
    country_df.groupby("Country")
    .size()
    .sort_values(ascending=False)
)

print(f"Number of country labels: {country_df['Country'].nunique()}")

display(country_counts.head(10))
display(country_counts.tail(10))

Number of country labels: 243


Country
Åland                        3239
Saint Pierre And Miquelon    3239
Romania                      3239
Poland                       3239
Norway                       3239
Greece                       3239
Netherlands                  3239
Montenegro                   3239
Monaco                       3239
Moldova                      3239
dtype: int64

Country
Papua New Guinea                       1581
Palmyra Atoll                          1569
Kingman Reef                           1569
Kiribati                               1569
Federated States Of Micronesia         1427
Northern Mariana Islands               1329
Guam                                   1329
Heard Island And Mcdonald Islands       788
French Southern And Antarctic Lands     788
Antarctica                              764
dtype: int64

## Initial Data-Quality Findings

- The global dataset contains **3,192 rows** and **9 columns**.
- The country dataset contains **577,462 rows** and **4 columns**.
- The global dataset covers **1750-01-01** to **2015-12-01**.
- The country dataset covers **1743-11-01** to **2013-09-01**.
- The most important missing-value findings were:
  **1,200 values (37.59%) are missing from each of the global maximum,
  minimum, and land-and-ocean temperature columns and their corresponding
  uncertainty columns. These variables contain no observations before
  January 1850. The global land-average temperature and uncertainty columns
  each contain 12 missing values (0.38%). In the country dataset,
  AverageTemperature contains 32,651 missing values (5.65%), while
  AverageTemperatureUncertainty contains 31,912 missing values (5.53%).
  Neither dataset contains missing dates, and the Country column contains
  no missing values.**
- The duplicated-record checks found:
  **0 duplicated dates in the global dataset and 0 duplicated
  country-and-date combinations in the country dataset.**
- The country field contains **243 distinct labels**.

These findings will be used to design the cleaning strategy. Missing
temperatures will not be filled automatically until their location and
frequency have been investigated.